File: JavaCheckWork1/var1/Stack.java
package var1;

public class Stack {
    Elem head; // Головной элемент стека

    // Конструктор класса Stack
    public Stack() {
        this.head = null;
    }

    // Метод для добавления элемента в стек
    public void push(Elem el) {
        if (el != null) {
            el.next = head; // Новый элемент указывает на текущий головной элемент
            head = el; // Головной элемент становится новым элементом
        }
    }

    // Метод для извлечения элемента из стека
    public Elem pop() {
        if (head == null) {
            return null; // Стек пуст, возвращаем null
        } else {
            Elem top = head; // Сохраняем верхний элемент стека
            head = top.next; // Головной элемент становится следующим элементом
            top.next = null; // Удаляем связь верхнего элемента
            return top; // Возвращаем извлеченный элемент
        }
    }
}

File: Алгоритмы и структуры данных/5/5/Stack.cpp
#include <iostream>
#include "Stack.h"

Stack::Stack() {
    head = NULL;
}

void Stack::add(int val) {
    Element* el = new Element;
    el->value = val;
    el->next = head;
    head = el;
}

bool Stack::isEmpty() {
    return head == NULL;
}

int Stack::getHead() {
    return head->value;
}

int Stack::deleteStart() {
    int val = head->value;
    Element* el = head;
    head = head->next;
    delete el;
    return val;
}

void Stack::print() {
    Element* el = head;
    while (el != NULL) {
        std::cout << el->value << " ";
        el = el->next;
    }
    std::cout << std::endl;
}

int Stack::size() {
    int c = 0;
    Element* el = head;
    while (el != NULL) {
        c++;
        el = el->next;
    }
    return c;
}

Stack::~Stack() {
    Element* el = head;
    Element* el1 = head;
    while (el != NULL) {
        el1 = el;
        el = el->next;
        delete el1;
    }
}
